### DISCUSSION TITLE
인터럽트는 무엇이고, 인터럽트의 종류와 처리과정에 대해 설명하세요.

### DISCUSSION THREAD

**Discussion by eatingbug** [Team-Psyduck/tech-sharing#61] [General]
추가질문
- 인터럽트와 폴링 방식의 차이점은 무엇인가요?
- 인터럽트는 어떤 상황에서 유용하나요?
- 하드웨어 인터럽트와 소프트웨어 인터럽트의 차이점은 무엇인가요?
- 인터럽트 핸들러는 어떻게 작동하나요?
- 운영체제에서 인터럽트 우선순위를 어떻게 관리하나요?
- 인터럽트와 멀티태스킹의 관계를 설명해주세요.
- 인터럽트에 대한 이해를 바탕으로 어떤 프로젝트에 적용해 본 경험이 있나요? 그 경험에 대해 설명해주세요.

**Comment by kkg5** [comment]
## 인터럽트

입출력이나 예외가 발생했을 때 CPU에 이를 알려 처리할 수 있도록 하는 것을 말합니다.

## 인터럽트 종류

- 외부 인터럽트

  - 전원 이상
  - 기계 고장
  - 입출력

- 내부 인터럽트

  - 프로그램 상의 오류

**Reply by SionBackEnd** [reply]
@kkg5 프로그램 상의 오류는 어떤 것들이 있을까요?

**Reply by kkg5** [reply]
`Division by zero`, `Overflow/Underflow`, `기타 프로그램 Exception`이 있습니다ㅎㅎ

[출처](https://itwiki.kr/w/%EC%9D%B8%ED%84%B0%EB%9F%BD%ED%8A%B8#%EB%82%B4%EB%B6%80_%EC%9D%B8%ED%84%B0%EB%9F%BD%ED%8A%B8)

**Comment by eatingbug** [comment]
인터럽트 (interrupt) 란 CPU가 프로그램을 실행하고 있을 때, 예외상황이 발생하여 처리가 필요한 경우 CPU에 알려 처리할 수 있도록 하는 것을 말한다.

인터럽트의 종류로는 하드웨어 인터럽트(interrupt)와 소프트웨어 인터럽트(trap) 이 있다. 통상적으로 인터럽트라고 하면 하드웨어 인터럽트를 의미하고, 소프트웨어 인터럽트는 트랩이라는 용어로 불린다. 소프트웨어 인터럽트의 예로는 예외상항(exception)과 시스템 콜(system call)이 있다.

폴링은 대상을 주기적으로 감시하여 상황이 발생하면 해당처리 루틴을 실행해 처리하는 반면, 인터럽트는 상대가 CPU에게 일을 처리해 달라고 요청하는 수단이다.

**Comment by SionBackEnd** [comment]
인터럽트란 CPU 사용 제어권을 운영체제에게 넘기는 것이다. 

인터럽트는 넓은 의미와 좁은 의미를 지니고 있다. 
- 넓은 의미 : 하드웨어가 발생시킨 인터럽트(키보드, 마우스)
- 좁은 의미 : Trap이라고 하며, 소프트웨어가 발생시킨 인터럽트 (Exception(timer interrupt), System call)

인터럽트 처리과정
인터럽트가 발생하면, 인터럽트 라인에 해당 인터럽트가 저장되고, CPU는 인터럽트 라인에 저장된 것을 매 작업마다 확인하며 처리한다. 
처리하기 위해서는 ProgramCounter 위치가 운영체제로 변경되고, 운영체제의 특권 명령을 이용하여 인터럽트를 처리한다.

**Reply by eatingbug** [reply]
`인터럽트가 발생하면, 인터럽트 라인에 해당 인터럽트가 저장되고, CPU는 인터럽트 라인에 저장된 것을 매 작업마다 확인하며 처리한다.`
매 작업마다 확인하며 처리하는 방식은 폴링 아닌가요?

[폴링(polling)이란 하나의 장치(또는 프로그램)가 충돌 회피 또는 동기화 처리 등을 목적으로 다른 장치(또는 프로그램)의 상태를 주기적으로 검사하여 일정한 조건을 만족할 때 송수신 등의 자료처리를 하는 방식을 말한다.](https://ko.wikipedia.org/wiki/%ED%8F%B4%EB%A7%81_(%EC%BB%B4%ED%93%A8%ED%84%B0_%EA%B3%BC%ED%95%99))

**Reply by SionBackEnd** [reply]
@Grooomit 
폴링과 인터럽트는 다른 방식 입니다. 
폴링 : **특정 주기**마다 스레드를 돌면서 시그널이 들어왔는지 확인하는 방식
인터럽트 : 외부 인터럽트 핀(인터럽트 라인)에 신호가 들어오면 CPU는 자신의 작업을 마무리하고 인터럽트 소스를 실행 후 다시 원래의 코드로 돌아오는 방식입니다. 

결국 폴링은 소프트웨어적인 방식으로 시그널을 확인하는 방법이고 인터럽트는 하드웨어 적인 방법으로 시그널을 확인하는 방식입니다. 

폴링은 구현이 쉬우나 시그널 발생 여부와 상관없이 특정 주기마다 확인해야 하기 때문에 리소스를 많이 먹고, 인터럽트는 구현이 복잡한 대신 인터럽트 발생 시에만 처리하면 되기 때문에 시스템 부하가 적습니다.

참조 링크 : https://kkhipp.tistory.com/155

**Reply by SionBackEnd** [reply]
`인터럽트가 발생하면, 인터럽트 라인에 해당 인터럽트가 저장되고, CPU는 인터럽트 라인에 저장된 것을 매 작업마다 확인하며 처리한다.`
아 방금 다시 확인해보니 제가 글을 잘못 적었네요 매 작업마다가 아니라 인터럽트 라인에 인터럽트가 존재할 때 만 확인하게 됩니다.

**Reply by eatingbug** [reply]
ㅋㅋㅋㅋ 정확한 답이 궁금해서 반효경 교수님 책을 찾아봤는데 이렇게 적혀있네요. 시온님이 처음 작성하신게 맞는 것 같습니다.

> CPU는 명령 하나를 수행할 때마다 인터럽트가 발생했는지 확인한다. 인터럽트가 발생했으면 다음 명령을 수행하기 전에 인터럽트 처리를 하게 되고, 그렇지 않으면 다음 명령을 계속 수행하게 되는 것이다. [운영체제와 정보기술의 원리 p70]
